Migrate `rustc_lint` lint diagnostics

rejyr opened the pull request:
Part 2 of [Migrate `rustc_lint` errors to `SessionDiagnostic`](https://github.com/rust-lang/rust/pull/100776)

r? @davidtwco 

# TODO
- [x] Refactor some lints manually implementing `DecorateLint` to use `Option<Subdiagnostic>`.
- [x] Add `#[rustc_lint_diagnostics]` to lint functions in `context.rs`.
- [x] Migrate `hidden_unicode_codepoints.rs`.
- [x] Migrate `UnsafeCode` in `builtin.rs`.
- [x] Migrate the rest of `builtin.rs`.

rejyr committed:
add: `lints.rs`

add: `lints.rs`
refactor: move `InvalidAtomicOrderingDiag` to `lints.rs`

rejyr committed:
migrate: `types.rs`

Rejyr commented:
@rustbot label: +A-diagnostics

Rejyr commented on compiler/rustc_lint/src/lints.rs:
```diff
@@ -0,0 +1,51 @@
+// FIXME: `#[derive(LintDiagnostic)] with #[suggestion_*]` does not fully output the suggestion,
+// bug maybe?
+// expected: ^^^^^^ help: use an inclusive range instead: `<suggestion>`
+// actual  : ^^^^^^ help: use an inclusive range instead
```
I'm having an issue with `#[derive(LintDiagnostic)]` with `#[suggestion_short]` where it doesn't show the replacement.

davidtwco replied:
If using suggestion, as in your manual implementation below, works, then just use #[suggestion(..)] here rather than #[suggestion_short]?

Rejyr replied:
It works now, thanks.

davidtwco commented on compiler/rustc_lint/src/lints.rs:
```diff
@@ -0,0 +1,51 @@
+// #[diag(lint::range_endpoint_out_of_range)]
+pub struct RangeEndpointOutOfRange<'a> {
+    pub ty: &'a str,
+    // #[suggestion_short(code = "{start}..={literal}{suffix}", applicability = "machine-applicable")]
```
```suggestion
    // #[suggestion_short(code = "{start}..={literal}{suffix}", applicability = "machine-applicable")]
```

If using `suggestion`, as in your manual implementation below, works, then just use `#[suggestion(..)]` here rather than `#[suggestion_short]`?

Rejyr commented:
@davidtwco How would I untangle `emit_ffi_unsafe_type_lint` and `FfiResult`? Would I only modify `emit_ffi_unsafe_type_lint`, create separate `LintDiagnostic` types for each lint, or just leave it?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8176920966?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/47bee6070144518b53072d4cc9f3e3d6d28667a6/checks/8176920966/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling fluent-syntax v0.11.0
   Compiling tracing v0.1.35
   Compiling tracing-subscriber v0.3.3
   Compiling thorin-dwp v0.3.0
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed
  |
  = note: delayed at    0: <rustc_errors::Handler>::delay_good_path_bug::<&str>
             1: rustc_middle::ty::print::pretty::trimmed_def_paths
             2: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::ArenaCache<(), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>>
             3: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::trimmed_def_paths, rustc_query_impl::plumbing::QueryCtxt>
             4: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::trimmed_def_paths
             5: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::Printer>::print_def_path
             7: <rustc_middle::ty::context::TyCtxt>::def_path_str
             7: <rustc_middle::ty::context::TyCtxt>::def_path_str
             8: <rustc_lint::traits::DropTraitConstraints as rustc_lint::passes::LateLintPass>::check_item
             9: <rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            10: <rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass> as rustc_hir::intravisit::Visitor>::visit_stmt
            11: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass>>
            12: <rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass> as rustc_hir::intravisit::Visitor>::visit_expr
            13: <rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_body
            14: <rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            15: rustc_hir::intravisit::walk_mod::<rustc_lint::late::LateContextAndPass<rustc_lint::BuiltinCombinedModuleLateLintPass>>
            16: rustc_lint::late::late_lint_mod::<rustc_lint::BuiltinCombinedModuleLateLintPass>
            18: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, ()>>
            19: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::lint_mod, rustc_query_impl::plumbing::QueryCtxt>
            20: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::lint_mod
            20: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::lint_mod
            21: <core::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures::sync::par_for_each_in<&[rustc_span::def_id::LocalDefId], <rustc_middle::hir::map::Map>::par_for_each_module<rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}> as core::ops::function::FnOnce<()>>::call_once
            22: rustc_data_structures::sync::par_for_each_in::<&[rustc_span::def_id::LocalDefId], <rustc_middle::hir::map::Map>::par_for_each_module<rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>
            23: <rustc_session::session::Session>::time::<(), rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#1}::{closure#0}>
            24: <core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}> as core::ops::function::FnOnce<()>>::call_once
            25: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}>, ()>
            26: <core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}> as core::ops::function::FnOnce<()>>::call_once
            27: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}>, ()>
            28: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}>
            30: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
            31: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
            32: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
            33: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            33: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            34: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
            35: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
            36: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>
            37: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            38: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            39: std::panicking::try::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, core::panic::unwind_safe::AssertUnwindSafe<<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
            40: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            41: std::sys::unix::thread::Thread::new::thread_start
            42: <unknown>
            43: <unknown>

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1499:13
stack backtrace:
   0:     0x7f5984e98ecd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h7aceaaec384abd8d
   0:     0x7f5984e98ecd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h7aceaaec384abd8d
   1:     0x7f5984efe9a8 - core::fmt::write::h1935094ec1b611e9
   2:     0x7f5984e89fe1 - std::io::Write::write_fmt::h924a08855ab35277
   3:     0x7f5984e9beee - std::panicking::default_hook::{{closure}}::hdcba8bc567463877
   4:     0x7f5984e9bbb7 - std::panicking::default_hook::h5e023012d652ad1c
   5:     0x7f598584ab04 - rustc_driver[491383e5761cfd5f]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7f5984e9c691 - std::panicking::rust_panic_with_hook::hc50598e1e398727d
   7:     0x7f59885cc163 - std[8a3c335779a4ef7b]::panicking::begin_panic::<rustc_errors[a9e596aa33e6456]::ExplicitBug>::{closure#0}
   8:     0x7f59885c91c6 - std[8a3c335779a4ef7b]::sys_common::backtrace::__rust_end_short_backtrace::<std[8a3c335779a4ef7b]::panicking::begin_panic<rustc_errors[a9e596aa33e6456]::ExplicitBug>::{closure#0}, !>
   9:     0x7f598580a9b6 - std[8a3c335779a4ef7b]::panicking::begin_panic::<rustc_errors[a9e596aa33e6456]::ExplicitBug>
  10:     0x7f59885bd666 - std[8a3c335779a4ef7b]::panic::panic_any::<rustc_errors[a9e596aa33e6456]::ExplicitBug>
  11:     0x7f59885c28ad - <rustc_errors[a9e596aa33e6456]::HandlerInner as core[c1e30f1bd259d119]::ops::drop::Drop>::drop
  12:     0x7f59858a6ae2 - core[c1e30f1bd259d119]::ptr::drop_in_place::<rustc_session[5e51afea9c8b884e]::parse::ParseSess>
  13:     0x7f59858aee15 - <alloc[45623a189840f9f9]::rc::Rc<rustc_session[5e51afea9c8b884e]::session::Session> as core[c1e30f1bd259d119]::ops::drop::Drop>::drop
  14:     0x7f598583b87c - core[c1e30f1bd259d119]::ptr::drop_in_place::<rustc_interface[62be15c45d0bd73b]::interface::Compiler>
  15:     0x7f5985839537 - rustc_span[8c7477ded0a91ee5]::with_source_map::<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_interface[62be15c45d0bd73b]::interface::create_compiler_and_run<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>::{closure#1}>
  16:     0x7f598584d98e - rustc_interface[62be15c45d0bd73b]::interface::create_compiler_and_run::<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>
  17:     0x7f59858304f2 - <scoped_tls[5efd78fa53ce51fd]::ScopedKey<rustc_span[8c7477ded0a91ee5]::SessionGlobals>>::set::<rustc_interface[62be15c45d0bd73b]::interface::run_compiler<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>
  18:     0x7f59858ab0bf - std[8a3c335779a4ef7b]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[62be15c45d0bd73b]::util::run_in_thread_pool_with_globals<rustc_interface[62be15c45d0bd73b]::interface::run_compiler<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>
  19:     0x7f598583abde - std[8a3c335779a4ef7b]::panicking::try::<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, core[c1e30f1bd259d119]::panic::unwind_safe::AssertUnwindSafe<<std[8a3c335779a4ef7b]::thread::Builder>::spawn_unchecked_<rustc_interface[62be15c45d0bd73b]::util::run_in_thread_pool_with_globals<rustc_interface[62be15c45d0bd73b]::interface::run_compiler<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  20:     0x7f59858ac8d0 - <<std[8a3c335779a4ef7b]::thread::Builder>::spawn_unchecked_<rustc_interface[62be15c45d0bd73b]::util::run_in_thread_pool_with_globals<rustc_interface[62be15c45d0bd73b]::interface::run_compiler<core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>, rustc_driver[491383e5761cfd5f]::run_compiler::{closure#1}>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>::{closure#0}, core[c1e30f1bd259d119]::result::Result<(), rustc_errors[a9e596aa33e6456]::ErrorGuaranteed>>::{closure#1} as core[c1e30f1bd259d119]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  21:     0x7f5984ea8665 - std::sys::unix::thread::Thread::new::thread_start::h3a069647ae68a533
  22:     0x7f5984c47b43 - <unknown>
  23:     0x7f5984cd9a00 - <unknown>
  24:                0x0 - <unknown>
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md


note: rustc 1.65.0-nightly (d6996aabb 2022-09-04) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C debuginfo=0 -C debug-assertions=on -Z unstable-options -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C split-debuginfo=off -Z unstable-options -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -Z binary-dep-depinfo -Z tls-model=initial-exec -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
```

</details>

rejyr committed:
migrate: `unused.rs`

davidtwco commented:
> @davidtwco How would I untangle `emit_ffi_unsafe_type_lint` and `FfiResult`? Would I only modify `emit_ffi_unsafe_type_lint`, create separate `LintDiagnostic` types for each lint, or just leave it?

I think `emit_ffi_unsafe_type_lint`/`FfiResult` might be okay as they are - this lint is using translated diagnostics already, it isn't in a type implementing `SessionDiagnostic` but that's probably okay for this case.

rejyr committed:
migrate: `traits.rs`

rejyr committed:
migrate: `redundant_semicolon.rs`

rejyr committed:
migrate: `pass_by_value.rs`

rejyr committed:
migrate: `noop_method_call.rs`

rejyr committed:
migrate: `nonstandard_style.rs`

rejyr committed:
migrate: `non_fmt_panic.rs`

Rejyr commented:
> @davidtwco How would I untangle `emit_ffi_unsafe_type_lint` and `FfiResult`? Would I only modify `emit_ffi_unsafe_type_lint`, create separate `LintDiagnostic` types for each lint, or just leave it?

@davidtwco Same thing for `check_panic` in `non_fmt_panic.rs`?

rejyr committed:
migrate: `non_ascii_idents.rs`

rejyr committed:
migrate: `methods.rs`

Rejyr commented:
@davidtwco How would I migrate usages of `struct_lint` and `struct_lint_level` in `levels.rs`?

davidtwco commented:
> @davidtwco How would I migrate usages of `struct_lint` and `struct_lint_level` in `levels.rs`?

We have `emit_spanned_lint` and `#[derive(LintDiagnostic)]` - I think that covers what you need for `struct_lint`? `struct_lint_level` might require a new function which does something similar to `emit_spanned_lint` - probably just changing its `decorate` argument to be a `impl DecorateLint`?

> @davidtwco Same thing for `check_panic` in `non_fmt_panic.rs`?

I think so, yeah.

Rejyr commented:
@davidtwco There's no `emit_spanned_lint` for `LintLevelsBuilder`, should I add one?  For `struct_lint_level`, in `levels.rs`, it only uses `LintLevelsBuilder.sess` for the `Session` and `LintLevelsBuilder.lint_level(..)` or `LintLevelsBuilder.provider.get_lint_level(..)` for the `Level` and `LintLevelSource`, so should I add a method like `emit_spanned_lint_level` to `LintLevelsBuilder`?

rejyr committed:
migrate: `array_into_iter.rs`

rejyr committed:
migrate: `early.rs` and `enum_intrinsics_non_enums.rs`

bors commented:
:umbrella: The latest upstream changes (presumably #101558) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rejyr committed:
add: `lints` for `errors.rs`

rejyr committed:
update: `lints.rs` for renamed traits

update: `lints.rs` for renamed `SessionSubdiagnostic` and `AddSubdiagnostic`

fix: NonSnakeCaseDiagSub

fix: OverflowingBinHexSign

davidtwco commented:
> @davidtwco There's no `emit_spanned_lint` for `LintLevelsBuilder`, should I add one? For `struct_lint_level`, in `levels.rs`, it only uses `LintLevelsBuilder.sess` for the `Session` and `LintLevelsBuilder.lint_level(..)` or `LintLevelsBuilder.provider.get_lint_level(..)` for the `Level` and `LintLevelSource`, so should I add a method like `emit_spanned_lint_level` to `LintLevelsBuilder`?

What diagnostic are you trying to migrate where this is necessary?

davidtwco reviewed (commented):
LGTM so far, thanks for your continued work on this.

Rejyr commented:
> > @davidtwco There's no `emit_spanned_lint` for `LintLevelsBuilder`, should I add one? For `struct_lint_level`, in `levels.rs`, it only uses `LintLevelsBuilder.sess` for the `Session` and `LintLevelsBuilder.lint_level(..)` or `LintLevelsBuilder.provider.get_lint_level(..)` for the `Level` and `LintLevelSource`, so should I add a method like `emit_spanned_lint_level` to `LintLevelsBuilder`?
> 
> What diagnostic are you trying to migrate where this is necessary?

@davidtwco Mostly the diagnostics in `levels.rs`. All of them are either `struct_lint` (one instance) or `struct_lint_level` (6 instances). 
There's a usage of `struct_lint` in `context.rs` at line 1013, but it's in the implementation of `LintContext` for `EarlyContext`.

Rejyr commented:
@davidtwco Also, should I make a new PR for refactoring the diagnostic structs in `errors.rs` since I've gotten more accustomed to migrating them, or can I add them to this PR?

davidtwco commented:
> @davidtwco Also, should I make a new PR for refactoring the diagnostic structs in `errors.rs` since I've gotten more accustomed to migrating them, or can I add them to this PR?

Feel free to include that in this pull request, but you can split it out if you want this to land first so it doesn't need rebasing as often.

davidtwco commented:
> @davidtwco Mostly the diagnostics in `levels.rs`. All of them are either `struct_lint` (one instance) or `struct_lint_level` (6 instances).
> There's a usage of `struct_lint` in `context.rs` at line 1013, but it's in the implementation of `LintContext` for `EarlyContext`.

I think you should add similar methods as `emit_spanned_lint` to `LintLevelsBuilder`, yeah.

bors commented:
:umbrella: The latest upstream changes (presumably #101986) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rejyr committed:
fix: merge conflict

rejyr committed:
refactor: comment about lint location

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3186317008/jobs/5196778825) [(plain)](https://github.com/rust-lang/rust/commit/9e20f49adfd08a3196e7c9d9cdeaff078b031017/checks/8711999903/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 855708bd0243
Successfully tagged rust-ci:latest
Built container sha256:855708bd02433eea7797c9d8fcd1f2f899b9aac0df3b9b95121bb05b8432df7e
Uploading finished image to https://ci-caches.rust-lang.org/docker/3ba9d538a45014cf9c069a7f7b39b17975213bb4ad3cb92953bcf42e2feac7246274891423ce6ec10459580375996bc5323b9b2e1eaa3f9fe83d7d3a2f2335cf
upload failed: - to s3://rust-lang-ci-sccache2/docker/3ba9d538a45014cf9c069a7f7b39b17975213bb4ad3cb92953bcf42e2feac7246274891423ce6ec10459580375996bc5323b9b2e1eaa3f9fe83d7d3a2f2335cf Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-llvm-13]
---
diff of stderr:

2   --> $DIR/lint-tool-test.rs:9:23
3    |
4 LL | #![cfg_attr(foo, warn(test_lint))]
-    |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
6    |
7    = note: `#[warn(renamed_and_removed_lints)]` on by default
7    = note: `#[warn(renamed_and_removed_lints)]` on by default
+ help: lint name `test_lint` is deprecated and may not have an effect in the future.
+    |
+ LL | #![cfg_attr(foo, warn(clippy::test_lint))]
8 
8 
9 warning: lint name `clippy_group` is deprecated and may not have an effect in the future.

11    |
11    |
12 LL | #![deny(clippy_group)]
-    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
+    |
+    |
+ help: lint name `clippy_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #![deny(clippy::group)]
14 
14 
15 warning: lint name `test_group` is deprecated and may not have an effect in the future.

17    |
17    |
18 LL | #[allow(test_group)]
-    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+    |
+    |
+ help: lint name `test_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #[allow(clippy::test_group)]
20 
21 warning: lint name `test_lint` is deprecated and may not have an effect in the future.
22   --> $DIR/lint-tool-test.rs:9:23


23    |
24 LL | #![cfg_attr(foo, warn(test_lint))]
-    |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
+    |
+    |
+ help: lint name `test_lint` is deprecated and may not have an effect in the future.
+    |
+ LL | #![cfg_attr(foo, warn(clippy::test_lint))]
26 
26 
27 warning: lint name `clippy_group` is deprecated and may not have an effect in the future.

29    |
29    |
30 LL | #![deny(clippy_group)]
-    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
+    |
+    |
+ help: lint name `clippy_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #![deny(clippy::group)]
32 
32 
33 warning: lint name `test_group` is deprecated and may not have an effect in the future.

35    |
35    |
36 LL | #[allow(test_group)]
-    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+    |
+    |
+ help: lint name `test_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #[allow(clippy::test_group)]
38 
39 warning: unknown lint: `this_lint_does_not_exist`
40   --> $DIR/lint-tool-test.rs:36:8


56   --> $DIR/lint-tool-test.rs:9:23
57    |
58 LL | #![cfg_attr(foo, warn(test_lint))]
-    |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
+    |
+    |
+ help: lint name `test_lint` is deprecated and may not have an effect in the future.
+    |
+ LL | #![cfg_attr(foo, warn(clippy::test_lint))]
60 
60 
61 warning: lint name `clippy_group` is deprecated and may not have an effect in the future.

63    |
63    |
64 LL | #![deny(clippy_group)]
-    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
+    |
+    |
+ help: lint name `clippy_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #![deny(clippy::group)]
66 
66 
67 error: item is named 'lintme'

89   --> $DIR/lint-tool-test.rs:31:9
90    |
90    |
91 LL | #[allow(test_group)]
-    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+    |
+    |
+ help: lint name `test_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #[allow(clippy::test_group)]
93 
94 warning: lint name `test_lint` is deprecated and may not have an effect in the future.
95   --> $DIR/lint-tool-test.rs:9:23


96    |
97 LL | #![cfg_attr(foo, warn(test_lint))]
-    |                       ^^^^^^^^^ help: change it to: `clippy::test_lint`
+    |
+    |
+ help: lint name `test_lint` is deprecated and may not have an effect in the future.
+    |
+ LL | #![cfg_attr(foo, warn(clippy::test_lint))]
99 
99 
100 warning: lint name `clippy_group` is deprecated and may not have an effect in the future.

102    |
102    |
103 LL | #![deny(clippy_group)]
-    |         ^^^^^^^^^^^^ help: change it to: `clippy::group`
+    |
+    |
+ help: lint name `clippy_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #![deny(clippy::group)]
105 
105 
106 warning: lint name `test_group` is deprecated and may not have an effect in the future.

108    |
108    |
109 LL | #[allow(test_group)]
-    |         ^^^^^^^^^^ help: change it to: `clippy::test_group`
+    |
+    |
+ help: lint name `test_group` is deprecated and may not have an effect in the future.
+    |
+ LL | #[allow(clippy::test_group)]
111 
112 error: aborting due to 2 previous errors; 14 warnings emitted
113 

---
To only update this specific test, also pass `--test-args lint-tool-test.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui-fulldeps/lint-tool-test.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--cfg" "foo" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui-fulldeps/lint-tool-test/auxiliary"
stdout: none
--- stderr -------------------------------
warning: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |
   = note: `#[warn(renamed_and_removed_lints)]` on by default
help: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(clippy::test_lint))]


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |
   |
help: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
LL | #![deny(clippy::group)]


warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |
   |
help: lint name `test_group` is deprecated and may not have an effect in the future.
   |
LL | #[allow(clippy::test_group)]

warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |
help: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(clippy::test_lint))]


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |
   |
help: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
LL | #![deny(clippy::group)]


warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |
   |
help: lint name `test_group` is deprecated and may not have an effect in the future.
   |
LL | #[allow(clippy::test_group)]

warning: unknown lint: `this_lint_does_not_exist`
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:36:8
   |
   |
LL | #[deny(this_lint_does_not_exist)] //~ WARNING unknown lint: `this_lint_does_not_exist`
   |
   = note: `#[warn(unknown_lints)]` on by default


warning: use of deprecated attribute `plugin`: compiler plugins are deprecated. See https://github.com/rust-lang/rust/pull/64675
   |
   |
LL | #![plugin(lint_tool_test)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: may be removed in a future compiler version
   = note: `#[warn(deprecated)]` on by default

warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |
help: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(clippy::test_lint))]


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |
   |
help: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
LL | #![deny(clippy::group)]


error: item is named 'lintme'
   |
   |
LL | fn lintme() { } //~ ERROR item is named 'lintme'
   |
note: the lint level is defined here
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:14:9
   |
   |
LL | #![deny(clippy_group)]
   |         ^^^^^^^^^^^^
   = note: `#[deny(clippy::test_lint)]` implied by `#[deny(clippy::group)]`

error: item is named 'lintmetoo'
   |
   |
LL |     fn lintmetoo() { } //~ ERROR item is named 'lintmetoo'
   |
   |
   = note: `#[deny(clippy::test_group)]` implied by `#[deny(clippy::group)]`

warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |
   |
help: lint name `test_group` is deprecated and may not have an effect in the future.
   |
LL | #[allow(clippy::test_group)]

warning: lint name `test_lint` is deprecated and may not have an effect in the future.
  --> /checkout/src/test/ui-fulldeps/lint-tool-test.rs:9:23
   |
   |
LL | #![cfg_attr(foo, warn(test_lint))]
   |
help: lint name `test_lint` is deprecated and may not have an effect in the future.
   |
   |
LL | #![cfg_attr(foo, warn(clippy::test_lint))]


warning: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #![deny(clippy_group)]
   |
   |
help: lint name `clippy_group` is deprecated and may not have an effect in the future.
   |
LL | #![deny(clippy::group)]


warning: lint name `test_group` is deprecated and may not have an effect in the future.
   |
   |
LL | #[allow(test_group)]
   |
   |
help: lint name `test_group` is deprecated and may not have an effect in the future.
   |
LL | #[allow(clippy::test_group)]

error: aborting due to 2 previous errors; 14 warnings emitted
------------------------------------------

```

</details>

rejyr committed:
add: `emit{,_spanned}_lint` for `LintLevelsBuilder`

add: `emit_spanned_lint` and `emit_lint` for `LintLevelsBuilder`
migrate: `DeprecatedLintName`

Rejyr commented on compiler/rustc_lint/src/levels.rs:
```diff
@@ -1081,6 +1072,18 @@ impl<'s, P: LintLevelsProvider> LintLevelsBuilder<'s, P> {
         struct_lint_level(self.sess, lint, level, src, span, msg, decorate)
     }
+
+    pub fn emit_lint(
```
@davidtwco

Rejyr replied:
There's also the option to have `emit_lint` and `emit_spanned_lint` as freestanding implementations on `LintLevelsBuilder`, but it would be better to unify them with `LintContext`.

davidtwco replied:
I'm a little confused, isn't there already an implementation of `emit_spanned_lint` and `emit_lint` on `LintContext`?

Rejyr replied:
There is, but `LintContext` is not implemented for `LintLevelsBuilder`. Implementing it for `LintLevelsBuilder` would require a type with trait `LintPassObject`, but only the traits `EarlyLintPass` and `LateLintPass` implement it.

davidtwco replied:
I'm not sure it's that much of a problem to just have it implemented like you have it now. You could add a new trait which has `emit_lint` and `emit_spanned_lint` and implement that for both `LintLevelsBuilder` and `LintContext` but I don't think we'd be gaining much.

rejyr committed:
migrate: `levels.rs`

rejyr committed:
migrate: `let_underscore.rs`

fix: NonBindingLetSub

rejyr committed:
migrate: `ImproperCTypes`

bors commented:
:umbrella: The latest upstream changes (presumably #102975) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Rejyr commented:
@rustbot label -T-infra -A-bootstrap -A-query-system

bors commented:
:umbrella: The latest upstream changes (presumably #103398) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rejyr committed:
migrate: `internal.rs`

rejyr committed:
migrate(wip): `builtin.rs`

Rejyr commented:
@davidtwco Is it fine to leave the other diagnostics in `builtin.rs` without Diagnostic structs? Also, the `rustc::diagnostic_outside_of_impl` and `rustc::untranslatable_diagnostic` don't seem to work since #101986.

Rejyr commented:
@rustbot label -A-query-system

bors commented:
:umbrella: The latest upstream changes (presumably #103426) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nilstrieb commented:
Hi, I've seen you changed some diagnostic structs in your PR. After #103345, the way we refer to fluent messages changed. They are now in a flat namespace with the same identifier as in the fluent file. For example,  `parser::cool_thing` is now `parser_cool_thing` and `parser::suggestion` just `suggestion`.
You should [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts) to the latest master and change your fluent message references as described above. Thanks!

Rejyr commented:
@rustbot label -A-query-system

Rejyr commented:
@rustbot label -A-query-system

bors commented:
:umbrella: The latest upstream changes (presumably #103727) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
> @davidtwco Is it fine to leave the other diagnostics in `builtin.rs` without Diagnostic structs? Also, the `rustc::diagnostic_outside_of_impl` and `rustc::untranslatable_diagnostic` don't seem to work since #101986.

Which diagnostics are those? Could you link some? I'm curious about the lints - the tests we have for them still work, maybe just some of the trigger functions had the annotation removed?

Rejyr commented:
> I'm curious about the lints - the tests we have for them still work, maybe just some of the trigger functions had the annotation removed?

@davidtwco The tests in [`diagnostics.rs`](https://github.com/rust-lang/rust/blob/534ddc6166a9031b0c269544929d68f2539ea7a0/src/test/ui-fulldeps/internal-lints/diagnostics.rs) only cover `Diagnostic` and the `err` functions, but not `LintDiagnostic` and the `lint` functions. 

Right now, `diagnostics.rs` has 
- `Diagnostic`
- `Handler::struct_err`
- `DiagnosticBuilder::note`. 

It's missing 
- `LintDiagnostic`
- `LintContext::{lookup_with_diagnostics,lookup,struct_span_lint,lint}`
- `TyCtxt::{struct_span_lint_hir,struct_lint_node}`
- `LintLevelsBuilder::struct_lint` 
- `struct_lint_level`. 

Essentially, all the functions changed to accept a `msg: impl Into<DiagnosticMessage>` or `msg: DiagnosticMessage`.


Since the `lint` functions were changed to accept a `msg`, there is no `lint.build(msg)` in the `decorate`, so `rustc::untranslatable_diagnostic` and `rustc::diagnostic_outside_of_impl` don't activate. Should those functions be linted against?

davidtwco commented:
> The tests in [`diagnostics.rs`](https://github.com/rust-lang/rust/blob/534ddc6166a9031b0c269544929d68f2539ea7a0/src/test/ui-fulldeps/internal-lints/diagnostics.rs) only cover `Diagnostic` and the `err` functions, but not `LintDiagnostic` and the `lint` functions.

@Rejyr I think that's mostly okay though - it tests that the mechanism works, if it works for one function it should work for all the others. For internal lints, I don't think we need to be too exhaustive about calling every single function that we trigger the lint for, it would probably be quite difficult to do that. Are you saying that the regular diagnostic functions, annotated with `#[rustc_lint_diagnostics]` so that the lint should trigger, aren't triggering the lint when they're in a closure being given to `struct_span_lint_hir`? Or are there just some functions missing a `#[rustc_lint_diagnostics]` annotation?

Rejyr commented:
> Are you saying that the regular diagnostic functions, annotated with `#[rustc_lint_diagnostics]` so that the lint should trigger, aren't triggering the lint when they're in a closure being given to `struct_span_lint_hir`? Or are there just some functions missing a `#[rustc_lint_diagnostics]` annotation?


@davidtwco Some functions are missing the `#[rustc_lint_diagnostics]` annotations.
- `LintContext::{lookup_with_diagnostics,lookup,struct_span_lint,lint}`
- `TyCtxt::struct_lint_node`
- `LintLevelsBuilder::struct_lint`.
I'll make a new PR for them.

Rejyr commented:
@rustbot -A-testsuite -T-infra

Rejyr commented:
@rustbot label -A-testsuite -T-infra

rejyr committed:
migrate: `expect.rs`

rejyr committed:
migrate: `for_loops_over_fallibles.rs`

rejyr committed:
refactor: refactor to derive for some lints.

rejyr committed:
add: `#[rustc_lint_diagnostics]` for more `context.rs` functions.

bors commented:
:umbrella: The latest upstream changes (presumably #104293) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rejyr committed:
migrate: `hidden_unicode_codepoints.rs`

rejyr committed:
migrate: `UnsafeCode` in `builtin.rs`

rejyr committed:
migrate: rest of `builtin.rs` without `builtin_asm_labels`

rejyr committed:
add: allow lints in `lints.rs`

bors commented:
:umbrella: The latest upstream changes (presumably #104600) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #104758) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
We can land this without the whole crate being complete if you'd prefer that.

Rejyr commented:
> We can land this without the whole crate being complete if you'd prefer that.

@davidtwco I'll fix the merge conflict, clean up the PR, and mark it as done. I'll migrate `BuiltinLintDiagnostics` and the usage of `lookup_with_diagnostics` in another PR.

rejyr committed:
refactor: cleanup

bors commented:
:umbrella: The latest upstream changes (presumably #103693) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rejyr committed:
migrate: `deref_into_dyn_supertrait.rs`

bors commented:
:umbrella: The latest upstream changes (presumably #105416) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #105883) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #106193) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #106324) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -2786,46 +2597,16 @@ impl<'tcx> LateLintPass<'tcx> for InvalidValue {
-                // FIXME(davidtwco): make translatable
-                cx.struct_span_lint(
+            if let Some(err) = with_no_trimmed_paths!(ty_find_init_error(cx, conjured_ty, init)) {
+                let init = match init {
```
Could these be made translatable?

Rejyr replied:
@davidtwco Yep, fixed it.

davidtwco reviewed (changes requested):
Apologies for the delay in reviewing this, r=me after fixing this last comment.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 0ea6b443f17d61d5fa494e012013856fe0ff4047 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 0ea6b443f17d61d5fa494e012013856fe0ff4047 -->
<!-- homu: {"type":"Approved","sha":"0ea6b443f17d61d5fa494e012013856fe0ff4047","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0ea6b443f17d61d5fa494e012013856fe0ff4047 with merge 196ca7c489d2cd2ee508930eb108ca7ff3def34e...
<!-- homu: {"type":"BuildStarted","head_sha":"0ea6b443f17d61d5fa494e012013856fe0ff4047","merge_sha":"196ca7c489d2cd2ee508930eb108ca7ff3def34e"} -->

rust-log-analyzer commented:
The job **`dist-apple-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3873309246/jobs/6603183878) [(plain)](https://github.com/rust-lang-ci/rust/commit/196ca7c489d2cd2ee508930eb108ca7ff3def34e/checks/10522879930/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] panic_unwind test:false 0.078
[RUSTC-TIMING] std_detect test:false 0.162
[RUSTC-TIMING] addr2line test:false 0.387
    Checking std v0.0.0 (/Users/runner/work/rust/rust/library/std)
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed but no error emitted; use `DelayDm` for lints or `with_no_trimmed_paths` for debugging
  = note: delayed at    0: std::backtrace_rs::backtrace::libunwind::trace
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
             1: std::backtrace_rs::backtrace::trace_unsynchronized
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
             2: std::backtrace::Backtrace::create
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/backtrace.rs:332:13
             3: std::backtrace::Backtrace::force_capture
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/backtrace.rs:314:9
             4: <rustc_errors::Handler>::delay_good_path_bug::<&str>
             5: rustc_middle::ty::print::pretty::trimmed_def_paths
             6: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::queries::trimmed_def_paths, rustc_query_impl::plumbing::QueryCtxt>
             7: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::trimmed_def_paths
             8: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::Printer>::print_def_path
             9: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::pretty_print_type
            10: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::pretty_print_type
            11: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::comma_sep::<rustc_middle::ty::Ty, core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::Ty>>>
            12: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::pretty_fn_sig
            13: <rustc_middle::ty::sty::FnSig as rustc_middle::ty::print::Print<rustc_middle::ty::print::pretty::FmtPrinter>>::print
            14: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::in_binder::<rustc_middle::ty::sty::FnSig>
            15: <rustc_middle::ty::sty::Binder<rustc_middle::ty::sty::FnSig> as core::fmt::Display>::fmt
            16: <rustc_lint::builtin::ClashingExternDeclarations as rustc_lint::passes::LateLintPass>::check_foreign_item
            17: <rustc_lint::late::RuntimeCombinedLateLintPass as rustc_lint::passes::LateLintPass>::check_foreign_item
            18: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_foreign_item
            19: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            20: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            21: rustc_hir::intravisit::walk_block::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            22: rustc_hir::intravisit::walk_expr::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            23: rustc_hir::intravisit::walk_body::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            24: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_body
            25: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            26: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            27: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            28: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            29: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            30: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            31: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            32: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            33: rustc_hir::intravisit::walk_item::<rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass>>
            34: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_item
            35: rustc_lint::late::late_lint_crate::<rustc_lint::BuiltinCombinedLateLintPass>
            36: <rustc_session::session::Session>::time::<(), rustc_lint::late::check_crate<rustc_lint::BuiltinCombinedLateLintPass, rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}::{closure#0}>::{closure#0}::{closure#0}>
            37: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}::{closure#1}::{closure#2}::{closure#0}>
            38: <core::panic::unwind_safe::AssertUnwindSafe<rustc_interface::passes::analysis::{closure#5}::{closure#1}> as core::ops::function::FnOnce<()>>::call_once
            39: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#5}>
            41: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
            42: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
            43: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#2}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            44: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
            44: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
            45: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            46: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            47: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            48: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/alloc/src/boxed.rs:1988:9
            49: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
            50: std::sys::unix::thread::Thread::new::thread_start
                       at /rustc/196ca7c489d2cd2ee508930eb108ca7ff3def34e/library/std/src/sys/unix/thread.rs:108:17
            51: __pthread_start
          
          

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.68.0-nightly (196ca7c48 2023-01-09) running on x86_64-apple-darwin

note: compiler flags: --crate-type dylib --crate-type rlib -C prefer-dynamic -C opt-level=3 -C embed-bitcode=no -C split-debuginfo=unpacked -C codegen-units=1 -C debuginfo=1 -Z unstable-options -C linker=/Users/runner/work/rust/rust/clang+llvm-14.0.5-x86_64-apple-darwin/bin/clang -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z macro-backtrace -Z osx-rpath-install-name -C link-args=-Wl,-rpath,@loader_path/../lib -C split-debuginfo=unpacked -C prefer-dynamic -C embed-bitcode=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
```

</details>

davidtwco commented:
@bors r+

bors commented:
:hourglass: Testing commit 88e5dd2530a48b046d35402807c4b3bb44775688 with merge c62631d39f8d2ecafc2bf57779be6bc835c116fd...
<!-- homu: {"type":"BuildStarted","head_sha":"88e5dd2530a48b046d35402807c4b3bb44775688","merge_sha":"c62631d39f8d2ecafc2bf57779be6bc835c116fd"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3882663018/jobs/6623516216)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3882663018/jobs/6623516216","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3882663018/jobs/6623079807) [(plain)](https://github.com/rust-lang-ci/rust/commit/c62631d39f8d2ecafc2bf57779be6bc835c116fd/checks/10548964980/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 88e5dd2530a48b046d35402807c4b3bb44775688 with merge bfffe406fbcabb37b95779f2d252c4a277191e0d...
<!-- homu: {"type":"BuildStarted","head_sha":"88e5dd2530a48b046d35402807c4b3bb44775688","merge_sha":"bfffe406fbcabb37b95779f2d252c4a277191e0d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3907751229/jobs/6678809997)
Approved by: davidtwco
Pushing bfffe406fbcabb37b95779f2d252c4a277191e0d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3907751229/jobs/6678809997"},"merge_sha":"bfffe406fbcabb37b95779f2d252c4a277191e0d"} -->

rust-timer commented:
Finished benchmarking commit (bfffe406fbcabb37b95779f2d252c4a277191e0d): [comparison URL](https://perf.rust-lang.org/compare.html?start=86ad69d4c63961b316fdbadd18bc35535f7b128b&end=bfffe406fbcabb37b95779f2d252c4a277191e0d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=86ad69d4c63961b316fdbadd18bc35535f7b128b&end=bfffe406fbcabb37b95779f2d252c4a277191e0d&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=86ad69d4c63961b316fdbadd18bc35535f7b128b&end=bfffe406fbcabb37b95779f2d252c4a277191e0d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.7%, 0.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.2%, 2.5%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.9% | [-3.2%, -2.7%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | -1.7% | [-3.2%, 0.7%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=86ad69d4c63961b316fdbadd18bc35535f7b128b&end=bfffe406fbcabb37b95779f2d252c4a277191e0d&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
